Add quality declaration ament_index_python (#50)
ament/ament_index: ament_index_python/Quality_Declaration.md

> 

Reviewer: ```suggestion
```

## Version Policy [1]

### Version Scheme [1.i]

`ament_index_python` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning)

### Version Stability [1.ii]

`ament_index_python` is not yet at a stable version, i.e. >= 1.0.0"

### Public API Declaration [1.iii]

Not defined API for this package

### API Stability Policy [1.iv]

`ament_index_python` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Policy [1.v]

`ament_index_python` is a Python package, so it does not need to handle ABI stability.

### ABI and ABI Stability Within a Released ROS Distribution [1.vi]

`ament_index_python` will not break API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

## Change Control Process [2]

`ament_index_python` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

### Change Requests [2.i]

> All changes will occur through a pull request, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

Reviewer: ```suggestion

All changes will occur through a pull request, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.
```

### Contributor Origin [2.ii]

> This package uses DCO as its confirmation of contributor origin policy. More information can be found in [CONTRIBUTING](./CONTRIBUTING.md)

Reviewer: ```suggestion

This package uses DCO as its confirmation of contributor origin policy. More information can be found in [CONTRIBUTING](./CONTRIBUTING.md)
```

### Peer Review Policy [2.iii]

> All pull request will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

Reviewer: ```suggestion

All pull request will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.
```

### Continuous Integration [2.iv]

All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)

###  Documentation Policy [2.v]

> All pull requests must resolve related documentation changes before merging

Reviewer: ```suggestion
All pull requests must resolve related documentation changes before merging.
```

## Documentation [3]

### Feature Documentation [3.i]

> `ament_index_python` does not have a documented feature list. Although it currently states part of its conceptual overview [here](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md)

Reviewer: ```suggestion
`ament_index_python` does not have a documented feature list. Although it currently states part of its conceptual overview [here](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md).
```

Currently nightly test results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)

## Dependencies [5]

`ament_index_python` has no run-time or build-time dependencies that need to be considered for this declaration.

## Platform Support [6]

`ament_index_python` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.

> 

Reviewer: Missing security section

> 

Reviewer 2: Remove empty lines

This document is a declaration of software quality for the `ament_index_python` package, based on the guidelines in [REP-2004](https://github.com/ros-infrastructure/rep/blob/rep-2004/rep-2004.rst).

# `ament_index_python` Quality Declaration

The package `ament_index_python` claims to be in the **Quality Level 4** category.

Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Quality Categories in REP-2004](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#package-quality-categories)

## Version Policy [1]

### Version Scheme [1.i]

`ament_index_python` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning)

### Version Stability [1.ii]

> Currently this package version is 0.8.0.''

Reviewer 2: Don't put specific versions as they change frequently, just say something like:

"`ament_index_python` is not yet at a stable version, i.e. >= 1.0.0"

### Public API Declaration [1.iii]

> Not defined API for this package

Reviewer 2: This does have a Python API, it's in the `ament_index_python` subdirectory. It's basically everything that's exported by `__all__` in https://github.com/ament/ament_index/blob/master/ament_index_python/ament_index_python/__init__.py

Author: I thought this was meant to be declared along with the "Features" documentation, stating it in the QD document will be valid IMO. However, what about the `cli.py` module? This is accessible as a package script

Reviewer 2: Going off one of the earlier examples, [rcutils](https://github.com/ros2/rcutils/blob/master/QUALITY_DECLARATION.md#public-api-declaration-1iii), the public API is kinda just declared in the QD. While the API is in the doxygen documentation, it's never explicitly stated anywhere else what the rcutils API is (as far as I can find). 

I guess I figured that was the point of section 1.iii, but maybe it's supposed be declared somewhere else?

Anything else that is intended to be used by packages outside of this package can probably be considered public API.

Author: The case of [rcutils](https://github.com/ros2/rcutils/blob/master/QUALITY_DECLARATION.md#public-api-declaration-1iii) is slightly different because `rcutils` is a C/C++ package, and for those the public headers are considered a valid definition of public API (according to [REP-2004](https://github.com/ros-infrastructure/rep/pull/218/files#diff-f40b01badf3f268cbb5d72ffcc9613f5R494), require a bit extra attention)

I think mentioning the API in a possible [README](./Readme.md) of this package will be enough to consider it a "declared API". Let me know your thoughts.

I have mixed feelings with the `cli` module as well, because their functions are not available but the script application is. I think it shouldn't be considered public API, and in that case I can add to the README file the rest of the functions.

Author: @dirk-thomas Can I have your thoughts on this as maintainer of the package?

Reviewer 3: I am sorry, but I don't have time for this atm.

### API Stability Policy [1.iv]

`ament_index_python` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Policy [1.v]

> `ament_index_python` must be concerned with ABI stability, and will maintain ABI stability within a ROS distribution.

Reviewer 2: ABI stability is a c concept, since this is python code it doesn't have to care about ABI stability.

### Continuous Integration [2.iv]

All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)

###  Documentation Policy [2.v]

All pull requests must resolve related documentation changes before merging

## Documentation [3]

### Feature Documentation [3.i]

> `ament_index_python` does not have a documented feature list. Although it currently states part of its concept overview [here]([https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md))

Reviewer 2: ```suggestion
`ament_index_python` does not have a documented feature list. Although it currently states part of its conceptual overview [here](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md)
```

### Public API Documentation [3.ii]

> `ament_index_python` does not cover a public API documentation.

Reviewer 2: Most of the public API functions are documented with docstrings in: https://github.com/ament/ament_index/blob/master/ament_index_python/ament_index_python/

But also state that it's not currently hosted anywhere.

Author: I'll amend this after the "public API declaration" is solved

## Testing [4]

### Feature Testing [4.i]

Each feature in `ament_index_python` has corresponding tests which simulate typical usage, and they are located in the [`test`](./test) directory.
New features are required to have tests before being added.
Currently nightly test results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)

New features are required to have tests before being added.

### Public API Testing [4.ii]

> As the API is not defined explicitly, this section is considered unresolved.

Reviewer 2: If we take the API to be the functions exported in `__init__.py`, then it looks like some of the functions are tested explicitly.

Author: Based on the [coverage report ](https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/ament_index_python/) the package tests all the functions of the package, so I changed the wording and result in the summary table

### Coverage [4.iii]

`ament_index_python` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#code-coverage), and opts to use line coverage instead of branch coverage.

This includes:

- tracking and reporting line coverage statistics
- achieving and maintaining a reasonable branch line coverage (90-100%)
- no lines are manually skipped in coverage calculations

Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.

> Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastBuild/cobertura/ament_index_python/)

Reviewer 2: Linking to the "lastBuild" of a ci_* type job is not stable. Link to the numbered job specifically:

https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/ament_index_python/

Author: But it will point to the closest being "Current", right? I think it's better having the latest, mind giving a third opinion? @ahcorde

Reviewer 2: The issue is that a ci_* type job is configured each time it's launched with whatever packages the person launching the job had wanted. The nightly_* jobs are different, since they automatically build with a fixed set of parameters. You can safely link to to the "lastBuild" on nightlies. Unfortunately, if I were to launch a new ci_linux_coverage job, but only for my own package, then your link becomes invalidated.

Reviewer 2: This is also only an issue with jobs linked to on ci.ros2.org. build.ros2.org has stable links

Author: You are right, I'll modify it to point to the numbered job.

### Performance [4.iv]

> `ament_index_python` follows the recommendations for performance testing of Python code in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#performance), and opts to do performance analysis on each release rather than each change.

Reviewer 2: Remove this sentence if there is no explicit performance testing.

Currently nightly test results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/ament_index_python.src.ament.ament_index.ament_index_python.test/)

## Dependencies [5]

`ament_index_python` has no run-time or build-time dependencies that need to be considered for this declaration.

## Platform Support [6]

`ament_index_python` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.

# Current status Summary

> The chart below compares the requirements in the REP-2004 with the current state of the rcutils package.

Reviewer 2: Update table with comments I made above. Also change `rcutils` to `ament_index_python`

### API Stability Policy [1.iv]

`ament_index_python` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Policy [1.v]

`ament_index_python` is a Python package, so it does not need to handle ABI stability.

### ABI and ABI Stability Within a Released ROS Distribution [1.vi]

`ament_index_python` will not break API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

## Change Control Process [2]

`ament_index_python` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

### Change Requests [2.i]

All changes will occur through a pull request, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

### Contributor Origin [2.ii]

This package uses DCO as its confirmation of contributor origin policy. More information can be found in [CONTRIBUTING](./CONTRIBUTING.md)

### Peer Review Policy [2.iii]

> All pull request will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

Reviewer 2: ```suggestion
All pull requests will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.
```

### Continuous Integration [2.iv]

> All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Reviewer 2: ```suggestion
All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```

### Public API Declaration [1.iii]

> `ament_index_python` public API is composed with the functions available in the module, exported in the [init file](./ament_index_python/__init__.py) and the [command line utility script](./ament_index_python/cli.py). 

Reviewer 2: ```suggestion
The public API of `ament_index_python` is composed of the functions available in the module, exported in the [init file](./ament_index_python/__init__.py) and the [command line utility script](./ament_index_python/cli.py). 
```

`ament_index_python` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

### Change Requests [2.i]

> All changes will occur through a pull request, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.

Reviewer 2: ```suggestion
All changes will occur through a pull request, check the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.
```

###  Documentation Policy [2.v]

All pull requests must resolve related documentation changes before merging.

## Documentation [3]

### Feature Documentation [3.i]

`ament_index_python` does not have a documented feature list. Although it currently states part of its conceptual overview [here](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md).

### Public API Documentation [3.ii]

> `ament_index_python` API is not documented.

Reviewer 2: Most of the functions have docstrings yah? It's just not hosted anywhere.

### Public API Testing [4.ii]

> All the functionality of this package is tested, having lines code coverage report result of [96%](https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/ament_index_python/).

Reviewer 2: Looking at the tests, I couldn't tell if each function was tested individually, or if they all were tested by the nature of the 'feature' tests. Do you know if each API function has associated tests?

Author: I can't tell for sure. But Public API Testing is considered as "Must have system, integration, and/or unit tests which cover all of the public API". So I think this shouldn't be a concern as long as the functions are covered.

Reviewer 2: Ok cool. The sentence could use a little cleanup though.

```suggestion
All the functionality of the declared API in this package is covered in its unit tests. Currently it has a line coverage of [96%](https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/ament_index_python/).
```
